rfd/0057-automatic-user-provisioning.md

```diff
@@ -0,0 +1,124 @@
+---
+authors: Alex McGrath (alex.mcgrath@goteleport.com)
+state: draft
+---
+
+# RFD 57 - Automatic user and sudoers provisioning
+
+## What
+
+Automatically create non-existing users and optionally add them to
+`sudoers` on Teleport nodes. Users will be removed after all sessions
+have logged out.
+
+## Why
+
+Currently, when logging into an SSH node, the user must be
+pre-created. Adding automatic user and `sudoer` provisioning would
+make it so that any Teleport user would be able to login and have the
+account created automatically without manual intervention.
+
+## Details
+
+The following are required for this feature:
+
+- Ability to automatically provision a Linux user if it's not present
+  on the node.
+- Ability to add the provisioned user to existing Linux groups defined
+  in the user traits/role.
+- Ability to add the provisioned user to sudoers.
+- Clean up the provisioned user / sudoers changes upon logout (being
+  careful not to remove pre-existing users).
+
+### Config/Role Changes
+
+Several new fields will need to be added to to the role `options` and
+`allow` sections:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: example
+spec:
+  options:
+    # Controls whether this role supports auto provisioning of users.
+    auto_create_user: true
+    # Controls whether this role allows adding auto provisioned users to sudoers.
+    add_to_sudoers: true
+  allow:
+    # New field listing Linux groups to assign a provisioned user to.
+    # Should support user and identity provider traits like other fields (e.g. "logins")
+    groups: [ubuntu, "{{internal.groups}}", "{{external.xxx}}"]
+```
+
+An individual `ssh_service` can be configured disable auto user
+creation with the below config:
+
+```yaml
+ssh_service:
+    # when disabled, takes precedence over the role setting
+    disable_auto_create_user: true
+```
+
+### User creation
+
+In order to create users `useradd` will be executed from teleport
+after a user has tried to access a Teleport SSH node.
+
+#### User Groups
+
+When a user is created they will be added to the specified groups from
+the role. In addition the user will be added to a special
+`teleport-system` group which can be used to indicate that the user
+was created by teleport and that its safe for it to be deleted. The
+groups will be created via `groupadd` at startup if they do not already
+exist and users will be added to groups via `usermod -aG <list of groups> <username>`
```

Comment:
I was originally just going to just create groups at startup and leave them around

I'm not sure how group deletion would work, it doesnt seem to cause an error if a group gets deleted while a user in that group has a running process. Maybe it could work by checking that there are no users in the teleport-system group anymore and then deleting groups made by teleport.

I think having a -teleport suffix would work but might make it a bit unclear when configuring software, why users have some groups with -teleport and some groups that already existed dont have a suffix,